### DISCUSSION TITLE
Asking for Feedback

### DISCUSSION THREAD

**Discussion by hugosenari** [cruel-intentions/devshell-files#3] [Q&A]
Welcome anyone!

Now that basic version are done we need community feedback.

Please leave your thoughts here.

**Comment by hugosenari** [comment]
@blaggacao sorry about spam, I thought you could share your wisdom here.

**Comment by blaggacao** [comment]
Thanks for letting me know. I'd have to wrap my head around this first.

This is a a code generator where the final artifact is (intended to be) committes.

This is interesting for use cases where two groups of people with disparate nix experience need to collaborate.

That is:
- nix expert produces config
- nix-noob needs to understand and audit it

However, being a code generator, it also requires pre-processing. Hence, another step is added. In gitops cycles that happen in short succession this can become quickly annoying, since at least when applying strict gitops, it requires an extra commit after the pre-proccessing step.

I'm currently in a similar situation where I need to audit & compare / semantic-diff final configuration artifacts, since diffing their build instructions written in nix would be utterly non-semantic (= ininteligible).

**Reply by hugosenari** [reply]
Thank you for your thoughts

To be honest I'm not fan of versioning generated stuff.
Sadly some cases require them, ie. how could I configure github actions as nix file without commit of generated file.
In the ideal world I all my nix configurations could be directly interpreted by my applications and tools,
To honest I'm not expecting it to be true as today I realized that [CFG](https://docs.red-dove.com/cfg/features.html#elements) is an old cool conf language that nobody uses.

Please, could you elaborate the audit part?
I mean, before decided use nix as config generator considered [lot of options](https://github.com/cruel-intentions/stylops#projects) (even the creation of my own solution), maybe [CUE](https://2021.splashcon.org/details/conflang-2021-papers/3/Large-Scale-Engineering-of-Configuration-with-Unification) better fit your case.

> ...ininteligible

Maybe we can separate workflow of `options` from `config`, as we can put them in separate files/directories/repositories, we could sanely diff theirs config instructions (env changes) instead of build instructions (validated at 'development time').

Curiously I liked nix because of its simplicity and fluent interface of object definition makes it looks more like a java properties
Unfortunately some configurations uses arrays even when order doesn't matters.

Ie:

```nix
{
  config.commands = [
    { package = "devshell.cli"; }
    { package = "convco"; }
  ];
  config.files.text = [ { name = "/hello.txt"; content = "Hello World!!"; } ];
  config.files.gitignore.enable = true;
  config.files.gitignore.patterns = [ { pattern = "*.txt"; enable = true; } ];
  config.files.gitignore.templates = [
     { template = "Global/Archives"; enable = true; }
     { template = "Global/Backup"; enable  = true; }
     { template = "Global/Diff"; enable  = true; }
     { 
        template = "VisualStudio"; enable  = false; 
     }
  ];
}
```

With lower cognitive load, even if those file get bigger, could be
```nix
{
  config.commands.package."convco" = "package description";
  config.commands.package."devshell.cli" = "packager description";
  config.files.text."/hello.txt" = "Hello World!!";
  config.files.gitignore.enable = true;
  config.files.gitignore.pattern."*.txt" = true;
  config.files.gitignore.template."Global/Archives" = true;
  config.files.gitignore.template."Global/Backup" = true;
  config.files.gitignore.template."Global/Diff" = true;
  config.files.gitignore.template."VisualStudio" = false;
}
```

It isn't possible today but give-me some ideas :smiling_imp:
- An alias of commands as dict
- An alias of imports as dict (I'm not sure if it works)
- An lexicographical sorted dict to replace arrays.

Again, thank you!
